### PR TITLE
Avalara Tax Reporting

### LINKED ISSUE #1704: Avalara tax reporting
Implement Avalara Tax reporting.  

Currently we've implemented rate look up, the next part of the process is reporting the rates when an order has been completed.  TaxCodes (assignment on the PDP, and submission during lookup) are also a requirement for implementation. 

Documentation: http://developer.avalara.com/avatax/

Related to #1692 and #1699.

### PR DESCRIPTION
**Pull request by zenweasel**
Closes #1704 

Uses TaxRate API to estimate taxes for cart
Record orders as "SalesInvoice" for Tax Reporting

Test plan:

With Avalara not enabled, ensure that you can checkout w/o issues (ensuring that address validation doesn't interfere)

1. Enable Avalara with key, username and password
1. Ensure that taxes are calculated correctly in cart
1. Complete an order
1. Login to admin panel (info in Wiki) and ensure that a "SalesInvoice" is recorded for that order
1. Refund that order
1. Ensure that a return is recorded

Testing Address validation:

1. Try a valid US address
1. Try an invalid US address
1. Try a valid Canadian Address (valid here means that the place actually exists)
1. Try an invalid Canadian address
1. Try an address from an unsupported country (PH is one)

### CODE REVIEW

File: imports/plugins/included/taxes-avalara/server/methods/taxCalc.js [hunk 1]
```diff
@@ -264,23 +264,26 @@ function orderToSalesInvoice(order) {
...
-  check(callback, Match.Optional(Function));
-
+  check(callback, Function);
   if (order.shipping && order.shipping[0].address) {
     const salesOrder = orderToSalesInvoice(order);
     const auth = getAuthData();
     const baseUrl = getUrl();
-    const requestUrl = `${baseUrl}/transactions/create`;
-    if (callback) {
+    const requestUrl = `${baseUrl}/transactions/creaaate`;
```
Comment (ltouro) [inline imports/plugins/included/taxes-avalara/server/methods/taxCalc.js]: there is a typo here, i think.

Comment (zenweasel) [inline imports/plugins/included/taxes-avalara/server/methods/taxCalc.js]: Thanks

Comment (aaronjudd) [inline imports/plugins/included/taxes-avalara/server/methods/taxCalc.js]: Looks like what happens when my cat helps code.

**Comment by kieckhafer** [comment]
Server and browser console errors when trying to add an address once avalara is on.

Server:
```
ERROR Reaction: failed [401] {"error":{"code":"AuthenticationException","message":"Authentication failed.","target":"HttpRequestHeaders","details":[{"code":"AuthenticationException","number":30,"message":"Authentication failed.","description":"Unable to authenticate the user or the account.","faultCode":"Client","helpLink":"http://developer.avalara.com/avatax/errors/AuthenticationException","severity":"Exception"}]}}
  Error: failed [401] {"error":{"code":"AuthenticationException","message":"Authentication failed.","target":"HttpRequestHeaders","details":[{"code":"AuthenticationException","number":30,"message":"Authentication failed.","description":"Unable to authenticate the user or the account.","faultCode":"Client","helpLink":"http://developer.avalara.com/avatax/errors/AuthenticationException","severity":"Exception"}]}}
      at Object.Future.wait (/Users/ek/.meteor/packages/meteor-tool/.1.4.2_3.1rd9djy++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/server-lib/node_modules/fibers/future.js:449:15)
      at Object.call (packages/meteor.js:213:24)
      at Object.HTTP.get (packages/http/httpcall_common.js:50:20)
      at Object.taxCalc.getCompanies (imports/plugins/included/taxes-avalara/server/methods/taxCalc.js:160:25)
      at Object.taxCalc.saveCompanyCode (imports/plugins/included/taxes-avalara/server/methods/taxCalc.js:170:31)
      at Object.taxCalc.getCompanyCode (imports/plugins/included/taxes-avalara/server/methods/taxCalc.js:86:36)
      at cartToSalesOrder (imports/plugins/included/taxes-avalara/server/methods/taxCalc.js:185:31)
      at Object.taxCalc.estimateCart (imports/plugins/included/taxes-avalara/server/methods/taxCalc.js:241:24)
      at [object Object].<anonymous> (imports/plugins/included/taxes-avalara/server/hooks/hooks.js:13:13)
      at server/api/method-hooks.js:122:36
      at arrayEach (/Users/ek/Projects/reaction/node_modules/lodash/_arrayEach.js:15:9)
      at forEach (/Users/ek/Projects/reaction/node_modules/lodash/forEach.js:38:10)
      at [object Object].MethodHooks._wrappers.(anonymous function) (server/api/method-hooks.js:121:5)
      at packages/check.js:130:16
      at [object Object]._.extend.withValue (packages/meteor.js:1122:17)
      at Object.exports.Match._failIfArgumentsAreNotAllChecked (packages/check.js:129:41)
      at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1709:18)
      at packages/ddp-server/livedata_server.js:1625:18
      at [object Object]._.extend.withValue (packages/meteor.js:1122:17)
      at Server.apply (packages/ddp-server/livedata_server.js:1624:45)
      at Server.call (packages/ddp-server/livedata_server.js:1567:17)
      at packages/kadira:flow-router-ssr/server/plugins/ssr_data.js:67:23
      - - - - -
      at makeErrorByStatus (packages/http/httpcall_common.js:13:10)
      at Request._callback (packages/http/httpcall_server.js:113:17)
      at Request.self.callback (/Users/ek/.meteor/packages/http/.1.2.10.3yb3pj++os+web.browser+web.cordova/npm/node_modules/request/request.js:200:22)
      at emitTwo (events.js:87:13)
      at Request.emit (events.js:172:7)
      at Request.<anonymous> (/Users/ek/.meteor/packages/http/.1.2.10.3yb3pj++os+web.browser+web.cordova/npm/node_modules/request/request.js:1067:10)
      at emitOne (events.js:82:20)
      at Request.emit (events.js:169:7)
      at IncomingMessage.<anonymous> (/Users/ek/.meteor/packages/http/.1.2.10.3yb3pj++os+web.browser+web.cordova/npm/node_modules/request/request.js:988:12)
      at emitNone (events.js:72:20)
      at IncomingMessage.emit (events.js:166:7)
      at endReadableNT (_stream_readable.js:923:12)
      at nextTickCallbackWith2Args (node.js:458:9)
      at process._tickCallback (node.js:372:17)
```


Browser: 
```
Error: Failed to add address: 
    at http://localhost:3000/app/app.js?hash=1a409c468d6bd6032cfc3a05e79bfd010d31a2a6:50266:27
    at MethodInvoker._callback (http://localhost:3000/packages/meteor.js?hash=e3f53db3be730057fed1a5f709ecd5fc7cae1229:1105:22)
    at MethodInvoker._maybeInvokeCallback (http://localhost:3000/packages/ddp-client.js?hash=bc32a166cd269e06a394f9418e0024d805bab379:3557:12)
    at MethodInvoker.dataVisible (http://localhost:3000/packages/ddp-client.js?hash=bc32a166cd269e06a394f9418e0024d805bab379:3586:10)
    at http://localhost:3000/packages/ddp-client.js?hash=bc32a166cd269e06a394f9418e0024d805bab379:4471:7
    at Array.forEach (native)
    at Function._.each._.forEach (http://localhost:3000/packages/underscore.js?hash=cde485f60699ff9aced3305f70189e39c665183c:149:11)
    at Connection._runAfterUpdateCallbacks (http://localhost:3000/packages/ddp-client.js?hash=bc32a166cd269e06a394f9418e0024d805bab379:4470:7)
    at Connection._performWrites (http://localhost:3000/packages/ddp-client.js?hash=bc32a166cd269e06a394f9418e0024d805bab379:4460:10)
    at Connection._flushBufferedWrites (http://localhost:3000/packages/ddp-client.js?hash=bc32a166cd269e06a394f9418e0024d805bab379:4423:10)
```

**Comment by zenweasel** [comment]
@kieckhafer You don't have the correct credentials in place. Please double-check that you have the key + username and password.

**Comment by kieckhafer** [comment]
Yep, I think that was it...  looks good now... i'm still seeing a warning in the browser console when the address is not validated, `Validation failed`, but the warning also appears in an alert and is only pre-validation. :+1:

**Pull request merged**